modules/vmfcore/src/compressor.cpp

```
@@ -0,0 +1,108 @@
+
+namespace vmf {
+
+typedef std::map< vmf_string, std::shared_ptr<Compressor> > CompressorsMap;
+enum CompressorType {USER=0, BUILTIN=1};
+
+CompressorsMap& getMapInstance(CompressorType type)
+{
+    //do that to prevent user from unregistering standard compressors
+    static CompressorsMap compressors[2];
+    if(compressors[BUILTIN].empty())
+    {
+        //register standard compressors
+        static std::shared_ptr<Compressor> dummy(std::make_shared<CompressorDummy>());
+        static std::shared_ptr<Compressor>  zlib(std::make_shared<CompressorZlib>());
+        compressors[BUILTIN][dummy->getId()] = dummy;
+        compressors[BUILTIN][ zlib->getId()] = zlib;
+    }
+    return compressors[type];
+}
+
+
+std::vector<vmf_string> Compressor::getRegisteredIds()
+{
+    std::vector<vmf_string> result;
+    for(CompressorType type: {USER, BUILTIN})
+    {
+        for(const auto& c : getMapInstance(type))
+        {
+            result.push_back(c.first);
+        }
+    }
+    return result;
+}
+
+
+void Compressor::registerNew(std::shared_ptr<Compressor> compressor)
+{
+    if(!compressor)
+    {
+        VMF_EXCEPTION(IncorrectParamException, "Incorrect instance of compressor");
+    }
+
+    if(create(compressor->getId()))
+    {
+        VMF_EXCEPTION(IncorrectParamException, "Compressor with that ID is already registered");
+    }
+
+    CompressorsMap& cmap = getMapInstance(USER);
+    cmap[compressor->getId()] = compressor;
+}
+
+
+std::shared_ptr<Compressor> Compressor::create(const vmf_string &id)
+{
+    CompressorsMap& userMap    = getMapInstance(USER);
+    CompressorsMap& builtinMap = getMapInstance(BUILTIN);
+
+    std::shared_ptr<Compressor> current;
+    if(builtinMap.find(id) != builtinMap.end())
+    {
+        current = builtinMap.at(id);
+    }
+    else if(userMap.find(id) != userMap.end())
+    {
+        current = userMap.at(id);
+    }
+
+    return current ? current->createNewInstance() : nullptr;
```

Let's throw an `IncorrectParamException` exception instead of `nullptr`